google/oauth2/service_account.py

```diff
@@ -436,6 +455,26 @@ def refresh(self, request):
                 "domain wide delegation is not supported for non-default universe domain"
             )
 
+        # If a refresh is triggered under the case of credential not expired
+        # that means some abnormal cases proactively trigger this refresh, we
+        # will attemp to fetch trust boundary but tolerant on errors if we
+        # already have cached trust boundaries.
```

Comment:
Can you expand on why the trust boundary cache needs to be refreshed?

Are trust boundaries expected to change periodically?

Reply:
Trust boundary won't be changed periodically. A one time set can last years long.
But during a runtime of a workload. Someone on the customer admin side can edit policies and cause a trust boundary change.
In this case, a request routing to a region which is no longer valid trust boundary will get `401` from backend.
And this `401` will naturally cause a refresh in most of the client as a common practice.